R/getSymbols.R

```diff
@@ -1428,6 +1428,136 @@ getSymbols.av <- function(Symbols, env, api.key,
 # Mnemonic alias, letting callers use getSymbols("IBM", src="alphavantage")
 getSymbols.alphavantage <- getSymbols.av
 
+#
+#  Download OHLC Data From Tiingo
+#  
+#  Meant to be called internally by getSymbols().
+#  
+getSymbols.tiingo <- function(Symbols, env, api.key,
+                              return.class="xts",
+                              periodicity="daily",
+                              adjusted = FALSE,
+                              from='2007-01-01',
+                              to=Sys.Date(),
+                              data.type="json",
+                              ...) {
+  assertArgs = function(arg, valid) {
+    if (!(eval(arg) %in% valid)) {
+      error_msg = paste0(arg, "must be a value of ",
+                         paste0(valid, collapse = ", "))
+      stop(error_msg)
+    }
+    invisible(NULL)
+  }
+  assertArgs(quote(return.class), c("xts", "data.frame", "zoo",
+                             "ts", "matrix", "timeSeries", "quantmod.OHLC"))
+  assertArgs(quote(periodicity), c("daily", "weekly", "monthly", "annually"))
+  assertArgs(quote(adjusted), c(TRUE, FALSE, "both"))
+  assertArgs(quote(data.type), c("json", "csv"))
```

Comment:
Josh we can throw this out if you don't like it. I wanted some form of argument assertion, though I usually use the checkmate package. The function above, such as the check on `adjusted`, returns an error if the argument value is not in the set. The error looks like
```
adjusted must be a value of TRUE, FALSE, both
```

Reply:
What's the benefit of this over `match.args()`?

Reply:
I reinvented the wheel here. Good catch will fix

Reply:
Ah, there is one edgecase
```r
match.arg(TRUE, c(TRUE, FALSE, "both"))
# Error in match.arg(TRUE, c(TRUE, FALSE, "both")) : 
#  'arg' must be NULL or a character vector
```
but I think I would rather have
```r
adjusted = match.arg(as.character(adjusted), c("TRUE", "FALSE", "both"))
# arg returns a char so we do conversion back to logical
if (adjusted == "TRUE" | adjusted = "FALSE") {
    adjusted = as.logical(adjusted)
}
```

or 

```r
_ = adjusted %in% c(TRUE, FALSE, "both") || 
    stop(paste0("Periodicity must a value of TRUE, FALSE, or 'both'"))
```

than my function. I'm going to do (1) unless you have a preference or alternative

Reply:
My preference is that `adjust` is logical only. No support for "both" because that's not provided by any other `getSymbols()` functions and I don't believe it's a common use case.

I would also prefer to take care of it myself.  I am amending some of your additions because they use different style/convention than the rest of the code (e.g. `=` for assignment, underscores in names, etc).  Some of your commits include changes from prior commits in master that you didn't have because your branch is behind upstream/master.  You also have merge commits, and I have a strong preference for rebasing versus including multiple merges of upstream/master into a feature branch.

Reply:
> My preference is that adjust is logical only. No support for "both" because that's not provided by any other getSymbols() functions and I don't believe it's a common use case.

Yeah I'm cool with that

> I would also prefer to take care of it myself.

Totes! If you need anything from me lmk. 

>  I am amending some of your additions because they use different style/convention than the rest of the code (e.g. = for assignment, underscores in names, etc). 

I'd rather not waste your time and would be happy to fix these things up myself.

> Some of your commits include changes from prior commits in master that you didn't have because your branch is behind upstream/master. You also have merge commits, and I have a strong preference for rebasing versus including multiple merges of upstream/master into a feature branch.

Yes I saw that on the contributor guide, rebased at first but then was waiting for you to look things over before I rebased pre-merge to master. I just rebased / squashed everything